### DISCUSSION TITLE
Understand alias on pyright

### DISCUSSION THREAD

**Discussion by Kludex** [pydantic/pydantic#3986] [General]
If I have this model:
```python
from pydantic import BaseModel, Field

class Potato(BaseModel):
    a: str = Field(..., alias="b")
    
    class Config:
        allow_population_by_field_name = True
```
The below will raise an issue on `pyright`:
```python
Potato(a="haha")
```
It will say that argument `b` is missing. I know about the dataclass transform thing, but is this behavior of `a` being ignored / cannot replace `b` expected? 🤔

**Comment by oefe** [comment]
This was added for Pedantic 1.9 in #2721 
Apparently, this doesn't consider aliasing – it isn't even mentioned in the PR or its discussion 
Particularly if `allow_population_by_field_name` is set to `True`, one will like to use the field names, not the aliases.

Maybe this could be changed in 1.9.1 to consider `allow_population_by_field_name`?

**Comment by samuelcolvin** [comment]
I suspect this well not be possible I think it's a fundamental limitation of pep 681.

**Reply by oefe** [reply]
Indeed, it seems this has already been discussed [here](https://github.com/microsoft/pyright/discussions/1782#discussioncomment-1920720) 😞

**Comment by squarebridges** [comment]
In case it's helpful for anyone else...

I've had a heck of a time with this problem since attempting to switch to Pyright (for the purposes of adopting Prefect, who both _love_ Pydantic, and only support Pyright). In one case (of many), I was using `allow_population_by_field_name = True` in a pydantic Model that represented an OpenAPI schema spec. OpenAPI schemata have lots of field names that have dashes in them, making them very difficult to represent in pydantic without using aliases. Snippet of an example:

openapi.yaml
```
x-google-endpoints:
- name: myservice.endpoints.myproject.cloud.goog
  target: 1.2.3.4
```

When using mypy, this was easily represented in the following Pydantic logic:
```py
class OpenAPISchema(BaseModel):
    x_google_endpoints: list[GoogleEndpoint] = Field([], alias="x-google-endpoints")

schema = OpenAPISchema(
    x_google_endpoints=[
        GoogleEndpoint(name="myservice.endpoints.myproject.cloud.goog", target="1.2.3.4")
    ],
)

assert schema.dict(by_alias=True)["x-google-endpoints"]["target"] == "1.2.3.4"
```

However, when switching to Pyright, I started getting the following two errors (as you've all also seen):

```
reportGeneralTypeIssues: Argument missing for parameter "x-google-endpoints"
reportGeneralTypeIssues: No parameter named "x_google_endpoints"
```

To get around this, I'm making use of `alias_generator` like so:

```py
from pydantic import BaseModel

ALIASES = {"x_google_endpoints": "x-google-endpoints"}


def alias_generator(name: str) -> str:
    return ALIASES.get(name, name)


class MyModel(BaseModel):
    x_google_endpoints: list[str]

    class Config:
        allow_population_by_field_name = True
        alias_generator = alias_generator


model = MyModel(x_google_endpoints=["my_endpoint"])

assert model.dict(by_alias=True)["x-google-endpoints"] == ["my_endpoint"]
assert MyModel.parse_obj({"x-google-endpoints": ["my_endpoint"]}).x_google_endpoints == [
    "my_endpoint"
]
```

Because pyright can't "see" the aliases, it still expects the parameter name to be the attribute name. And of course, the result of `.dict(by_alias=True)` isn't a TypedDict, just a regular `dict[str, Any]`, so there's no typing issues there. It's a bit of a refactor, but well worth being able to use Pyright!

At least until I find the next Pyright/Pydantic conflict. Hopefully that's not for a while.

**Reply by geekmorn** [reply]
Thanks man, you help me a lot ;)

**Comment by Ed1123** [comment]
Any update? The only workaround I've found is to ignore the error... (# type: ignore)

**Comment by jarmos-san** [comment]
As of Pydantic v2.10, the issue seems to be fixed! Passing `populate_by_name=True` to `ConfigDict()` seems to satisfy Pyright/Pylance (and mypy too).

Here is a working solution ([half of which was picked from the docs and modified](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.alias_generator)):

```python
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

class Voice(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    name: str
    language_code: str

voice = Voice(name='Filiz', language_code='tr-TR')
print(voice.language_code)
#> tr-TR
print(voice.model_dump(by_alias=True))
#> {'name': 'Filiz', 'languageCode': 'tr-TR'}
```

**Reply by Viicos** [reply]
You haven't defined any aliases on `name` and `language_code`, so this is working as expected. As [mentioned](https://github.com/pydantic/pydantic/discussions/3986#discussioncomment-2624741), this is a limitation of PEP 681, although in fact there's no way to reasonably support this in practice (see https://github.com/pydantic/pydantic/issues/10168#issuecomment-2598973423, https://github.com/pydantic/pydantic/issues/5893#issuecomment-2512807073).
